benchmarks/_sources.py

```diff
@@ -45,11 +44,60 @@ class SourceFormat(Enum):
     FEATHER = "feather"
 
 
+# This is a reconstructed schema; it may not be completely accurate, but should
+# be serviceably so. For more information on variables, see the sources below
+# available from Fannie Mae. You may need to create an account first here:
+# https://datadynamics.fanniemae.com/data-dynamics/#/reportMenu;category=HP
+#
+# Data dictionary: https://capitalmarkets.fanniemae.com/resources/file/credit-risk/xls/crt-file-layout-and-glossary.xlsx
+# Names from R code here: https://capitalmarkets.fanniemae.com/media/document/zip/FNMA_SF_Loan_Performance_r_Primary.zip
+fannie_mae_schema = pyarrow.schema(
+    [
+        pyarrow.field("LOAN_ID", pyarrow.string()),
+        pyarrow.field("ACT_PERIOD", pyarrow.string()),  # date. Monthly reporting period
+        pyarrow.field("SERVICER", pyarrow.string()),
+        pyarrow.field("ORIG_RATE", pyarrow.float64()),
+        pyarrow.field("CURRENT_UPB", pyarrow.float64()),
+        pyarrow.field("LOAN_AGE", pyarrow.int32()),
+        pyarrow.field("REM_MONTHS", pyarrow.int32()),
+        pyarrow.field("ADJ_REM_MONTHS", pyarrow.int32()),
+        pyarrow.field("MATR_DT", pyarrow.string()),  # maturity date
+        pyarrow.field("MSA", pyarrow.string()),  # Metropolitan Statistical Area code
+        pyarrow.field(
+            "DLQ_STATUS", pyarrow.string()
+        ),  # Int of months, but `X` is a valid value. New versions pad with `0`/`X` to two characters
+        pyarrow.field("RELOCATION_MORTGAGE_INDICATOR", pyarrow.string()),
+        pyarrow.field(
+            "Zero_Bal_Code", pyarrow.string()
+        ),  # 0-padded 2 digit ints representing categorical levels, e.g. "01" -> "Prepaid or Matured"
+        pyarrow.field("ZB_DTE", pyarrow.string()),  # date
+        pyarrow.field("LAST_PAID_INSTALLMENT_DATE", pyarrow.string()),
+        pyarrow.field("FORECLOSURE_DATE", pyarrow.string()),
+        pyarrow.field("DISPOSITION_DATE", pyarrow.string()),
+        pyarrow.field("FORECLOSURE_COSTS", pyarrow.float64()),
+        pyarrow.field("PROPERTY_PRESERVATION_AND_REPAIR_COSTS", pyarrow.float64()),
+        pyarrow.field("ASSET_RECOVERY_COSTS", pyarrow.float64()),
+        pyarrow.field("MISCELLANEOUS_HOLDING_EXPENSES_AND_CREDITS", pyarrow.float64()),
+        pyarrow.field("ASSOCIATED_TAXES_FOR_HOLDING_PROPERTY", pyarrow.float64()),
+        pyarrow.field("NET_SALES_PROCEEDS", pyarrow.float64()),
+        pyarrow.field("CREDIT_ENHANCEMENT_PROCEEDS", pyarrow.float64()),
+        pyarrow.field("REPURCHASES_MAKE_WHOLE_PROCEEDS", pyarrow.float64()),
+        pyarrow.field("OTHER_FORECLOSURE_PROCEEDS", pyarrow.float64()),
+        pyarrow.field("NON_INTEREST_BEARING_UPB", pyarrow.float64()),
+        pyarrow.field("MI_CANCEL_FLAG", pyarrow.string()),  # all null
+        pyarrow.field("RE_PROCS_FLAG", pyarrow.string()),
+        pyarrow.field("LOAN_HOLDBACK_INDICATOR", pyarrow.string()),  # all null
+        pyarrow.field("SERV_IND", pyarrow.string()),
+    ]
+)
```

Comment:
@joosthooz The above is my best attempt at a schema for `fanniemae_2016Q4`; I never figured out how it was subset, so I reconstructed what I could by hand with a couple guesses that should be close enough for our purposes. It's not perfect, but it's a lot better. Let me know if I should PR it in somewhere else, or feel free to grab this and insert it wherever you need.

cc @austin3dickey

Reply:
thanks!! We can put it [here](https://github.com/voltrondata/datasources/blob/installing/dcg/repo.json) eventually but that's on a branch right now so I'll do it later